### PR TITLE
chore: Make binary_mut kernel accept different type for second arg

### LINKED ISSUE #5818: `binary_mut` kernel requires both args to be the same type (which is inconsistent with `binary`)
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As part of documenting the `_mut` compute kernels better in https://github.com/apache/arrow-rs/pull/5798 I noticed in https://github.com/apache/arrow-rs/pull/5798#discussion_r1613555891 thta 

[binary_mut](https://docs.rs/arrow/latest/arrow/compute/fn.binary_mut.html) is inconsistent with [`binary`](https://docs.rs/arrow/latest/arrow/compute/fn.binary.html) as it requires both `a` and `b` to be the same type of array (where `binary` can take different types)

**Describe the solution you'd like**
I would like to have the argument types to `binary_mut` be different -- so for example, I could write a kernel using `binary_mut` that took a `Float32Array` and a `Int32Array` and added the int32 vale to the float in place. 

**Describe alternatives you've considered**
Leave as is

**Additional context**
@viirya I think added this API in https://github.com/apache/arrow-rs/pull/3144

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #5818.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
This is technically an API change as it could effect type inference

**Comment by viirya** [comment]
Thank you @tustvold

File: arrow-arith/src/arity.rs [hunk 1]
```diff
@@ -224,14 +224,15 @@ where
 /// This function gives error if the arrays have different lengths.
 /// This function gives error of original [`PrimitiveArray`] `a` if it is not a mutable
 /// primitive array.
-pub fn binary_mut<T, F>(
+pub fn binary_mut<T, U, F>(
```
Comment (alamb) [inline arrow-arith/src/arity.rs:227]: FWIW I have added a doc example (and thus test coverage) for this change in https://github.com/apache/arrow-rs/pull/5798

**Review by alamb — COMMENTED** [review]
Thanks @viirya

File: arrow-arith/src/arity.rs (hunk 1 above)
Comment (viirya) [reply arrow-arith/src/arity.rs:227]: Thank you @alamb

**Pull request merged**